File: python/tests/test_weighted.py
"""Test weighted regression functionality."""

import polars as pl
import numpy as np
from leanfe import leanfe_polars


def test_weighted_regression():
    """Test basic weighted regression."""
    np.random.seed(42)
    n = 1000
    
    x = np.random.randn(n)
    fe1 = np.random.randint(0, 10, n)
    weights = np.random.uniform(0.5, 2.0, n)
    y = 2.0 * x + np.random.randn(n)
    
    df = pl.DataFrame({
        "y": y,
        "x": x,
        "fe1": fe1,
        "weights": weights
    })
    
    result = leanfe_polars(df, "y", ["x"], ["fe1"], weights="weights")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]
    assert abs(result["coefficients"]["x"] - 2.0) < 0.2  # Close to true value


def test_weighted_with_formula():
    """Test weighted regression with formula API."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "weights": np.random.uniform(0.5, 2.0, n)
    })
    
    result = leanfe_polars(df, formula="y ~ x | fe1", weights="weights")
    
    assert "coefficients" in result
    assert "x" in result["coefficients"]


def test_weighted_with_robust_se():
    """Test weighted regression with HC1 standard errors."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        "y": np.random.randn(n),
        "x": np.random.randn(n),
        "fe1": np.random.randint(0, 10, n),
        "weights": np.random.uniform(0.5, 2.0, n)
    })
    
    result = leanfe_polars(df, "y", ["x"], ["fe1"], 
                               weights="weights", vcov="HC1")
    
    assert result["vcov_type"] == "HC1"
    assert "std_errors" in result

File: r/tests/test_weighted.R
#!/usr/bin/env Rscript
# Test weighted regression functionality

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/compress.R")
source("R/polars.R")
source("R/duckdb.R")
source("R/leanfe.R")

set.seed(42)
n <- 1000

cat("Testing weighted regression...\n\n")

# Test 1: Basic weighted regression
cat("Test 1: Basic weighted regression... ")
x <- rnorm(n)
fe1 <- sample(0:9, n, replace = TRUE)
weights <- runif(n, 0.5, 2.0)
y <- 2.0 * x + rnorm(n)

df <- pl$DataFrame(
  y = y,
  x = x,
  fe1 = fe1,
  weights = weights
)

result <- leanfe_polars(df, y_col = "y", x_cols = "x", fe_cols = "fe1", weights = "weights")

stopifnot("coefficients" %in% names(result))
stopifnot("x" %in% names(result$coefficients))
stopifnot(abs(result$coefficients["x"] - 2.0) < 0.2)  # Close to true value
cat("PASSED\n")

# Test 2: Weighted regression with formula API
cat("Test 2: Weighted regression with formula... ")
df <- pl$DataFrame(
  y = rnorm(n),
  x = rnorm(n),
  fe1 = sample(0:9, n, replace = TRUE),
  weights = runif(n, 0.5, 2.0)
)

result <- leanfe_polars(df, formula = "y ~ x | fe1", weights = "weights")

stopifnot("coefficients" %in% names(result))
stopifnot("x" %in% names(result$coefficients))
cat("PASSED\n")

# Test 3: Weighted regression with HC1 standard errors
cat("Test 3: Weighted regression with HC1 SE... ")
result <- leanfe_polars(
  df, 
  y_col = "y", 
  x_cols = "x", 
  fe_cols = "fe1",
  weights = "weights", 
  vcov = "HC1"
)

stopifnot(result$vcov_type == "HC1")
stopifnot("std_errors" %in% names(result))
cat("PASSED\n")

cat("\n✅ All 3 tests passed!\n")
